File: api/src/middleware/cors.ts
import { config } from "../config";

export function corsMiddleware(
  req: { headers: { origin?: string }; method?: string },
  res: {
    setHeader: (key: string, value: string) => void;
    end: () => void;
    statusCode: number;
  },
  next: () => void,
) {
  const origin = req.headers.origin;

  if (!origin) {
    return next();
  }

  if (!config.cors.origins.includes(origin)) {
    res.statusCode = 403;
    return res.end();
  }

  res.setHeader("Access-Control-Allow-Origin", origin);
  res.setHeader("Vary", "Origin");
  res.setHeader("Access-Control-Allow-Methods", "GET,POST,PUT,PATCH,DELETE,OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type, Authorization");

  if (req.method === "OPTIONS") {
    res.statusCode = 204;
    return res.end();
  }

  next();
}

File: backend/internal/middleware/cors.go
package middleware

import (
	"bookearth/backend/internal/config"
	"net/http"
)

func CORS(next http.Handler) http.Handler {
	cfg := config.LoadConfig()
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		origin := r.Header.Get("Origin")
		if origin == "" {
			next.ServeHTTP(w, r)
			return
		}

		allowed := false
		for _, o := range cfg.CORS.Origins {
			if origin == o {
				allowed = true
				break
			}
		}

		if !allowed {
			w.WriteHeader(http.StatusForbidden)
			return
		}

		w.Header().Set("Access-Control-Allow-Origin", origin)
		w.Header().Set("Vary", "Origin")
		w.Header().Set("Access-Control-Allow-Methods", "GET,POST,PUT,PATCH,DELETE,OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusNoContent)
			return
		}

		next.ServeHTTP(w, r)
	})
}
